v3/mathjax3-ts/util/LinkedList.ts

```diff
@@ -0,0 +1,356 @@
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  Implement a generic LinkedList object.
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+/*****************************************************************/
+/*
+ *  A symbol used to mark the special node used to indicate
+ *  the start and end of the list.
+ */
+const END = Symbol();
+
+/*
+ * Shorthand type for the functions used to sort the data items
+ */
+export type SortFn<DataClass> = (a: DataClass, b: DataClass) => boolean;
+
+/*****************************************************************/
+/*
+ *  The ListItem interface (for a specific type of data item)
+ *
+ *  These are the items in the doubly-linked list.
+ */
+
+export class ListItem<DataClass> {
+    /*
+     * The data for the list item
+     */
+    public data: DataClass | symbol;
+
+    /*
+     * Pointers to the next and previous items in the list
+     */
+    public next: ListItem<DataClass> = null;
+    public prev: ListItem<DataClass> = null;
+
+    /*
+     * @param{any} data  The data to be stored in the list item
+     * @constructor
+     */
+    constructor (data: any = null) {
+        this.data = data;
+    }
+}
+
+/*****************************************************************/
+/*
+ *  The ListList class interface
+ */
+
+export interface LinkedListClass<DataClass> {
+    new(...args: DataClass[]): LinkedList<DataClass>;
+}
+
+/*****************************************************************/
+/*
+ *  Implements the generic ListList class
+ */
+
+export class LinkedList<DataClass> {
+
+    /*
+     * The linked list
+     */
+    protected list: ListItem<DataClass>;
+
+    /*
+     *  This.list is a special ListItem whose next property
+     *    points to the head of the list and whose prev
+     *    property points to the tail.  This lets us relink
+     *    the head and tail items in the same way as any other
+     *    item in the list, without having to handle special
+     *    cases.
+     *
+     * @param{DataClass[]} args  The data items that form the initial list
+     * @constructor
+     */
+    constructor(...args: DataClass[]) {
+        this.list = new ListItem<DataClass>(END);
+        this.list.next = this.list.prev = this.list;
+        this.push(...args);
+    }
+
+    /*
+     * Typescript targeted at ES5 doesn't handle
+     *
+     *     for (const x of this) {...}
+     *
+     * so use toArray() to convert to array, when needed
+     *
+     * @return{DataClass[]}  The list converted to an array
+     */
+    public toArray() {
+        return Array.from(this) as DataClass[];
+    }
+
+    /*
+     *  Used for sorting and merging lists (Overridden by subclasses)
+     *
+     * @param{DataClass} a   The first item to compare
+     * @param{DataClass} b   The second item to compare
+     * @return{boolean}      True if a is before b, false otherwise
+     */
+    public isBefore(a: DataClass, b: DataClass) {
+        return (a < b);
```

Comment:
Yes, the sorted list will have the same number of elements as the original, even if some are "equal".  There is nothing about the list that requires the elements to be unique.  If the two are equal, then their order doesn't really matter and their order (among those of the same value) is essentially random.  If there should be a distinction among them, then you would want to make your `isBefore()` function take that into account (e.g., `return  a.i < b.i || (a.i === b.i && a.j < b.j)`).